Bug: Pager error messages

### Description of the issue

**Description:**
Running `cargo nextest list` opens in a pager, but if you close the pager before reaching the end of the list, you get a weird error message.

**Steps to reproduce:**
1. Run `cargo nextest list` in a project with many tests (e.g. https://github.com/KittyCAD/modeling-app/, cd into `rust/` dir)
2. Hit "q" without scrolling


### Expected outcome

I should not see an error message.

### Actual result

This error message:


```
error: failed to write test list to output

Caused by:
  error writing to output

Caused by:
  Broken pipe (os error 32)
```

### Nextest version

```text
cargo-nextest 0.9.119-b.1 (05c3fdb33 2026-01-05)
release: 0.9.119-b.1
commit-hash: 05c3fdb334934d58e8417086298f9a9043fef500
commit-date: 2026-01-05
host: aarch64-apple-darwin
```

### Additional context

This doesn't happen if I scroll to the end before hitting "q".

Thanks for the report. I was unable to build that repo on macOS:

```
error[E0412]: cannot find type `PyEncodingWarning` in this scope
   --> /Users/rain/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/pyo3-stub-gen-0.13.1/src/exception.rs:69:27
    |
 46 |           impl crate::PyStubType for $name {
    |                                      ----- due to this macro variable
...
 69 |   impl_exception_stub_type!(PyEncodingWarning, "EncodingWarning");
    |                             ^^^^^^^^^^^^^^^^^ help: a struct with a similar name exists: `PyUnicodeWarning`
    |
   ::: /Users/rain/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/pyo3-0.25.1/src/exceptions.rs:725:1
    |
725 | / impl_native_exception!(
726 | |     PyUnicodeWarning,
727 | |     PyExc_UnicodeWarning,
728 | |     native_doc!("UnicodeWarning")
729 | | );
    | |_- similarly named struct `PyUnicodeWarning` defined here
```

But I'd imagine there isn't much repo-specific configuration involved here, and I've not been able to reproduce it against other repos and on macOS/Linux/illumos.

Some questions:

* What version of macOS are you on?
* What does `less --version` say for you?
* Do you have LESSFLAGS set in your environment?
* Could you give it a try on another platform like Linux, if you have access to it?
* While nextest is running, could you grab the `less` invocation from `ps` output? `ps -ef | grep less` should show something like `502 66053 66033   0  1:21PM ttys000    0:00.00 less -FRX`.

Thanks!

--- a/cargo-nextest/src/dispatch/clap_error.rs
+++ b/cargo-nextest/src/dispatch/clap_error.rs
@@ -15,6 +15,7 @@ use nextest_runner::{
     pager::PagedOutput,
     platform::Platform,
     user_config::{EarlyUserConfig, elements::PaginateSetting},
+    write_str::WriteStr,
 };
 use std::io::{IsTerminal, Write};
 use tracing::debug;
@@ -198,11 +199,17 @@ fn handle_help_output(err: clap::Error, early_args: &EarlyArgs, host_platform: &
         &early_config.streampager,
     );
 
-    if let Err(e) = paged.stdout().write_all(help_text.as_bytes()) {
+    if let Err(error) = paged.write_str(&help_text) {
         // If writing to pager fails, try stdout directly.
-        debug!("failed to write to pager: {e}, falling back to stdout");
+        debug!("failed to write to pager: {error}, falling back to stdout");
         let _ = std::io::stdout().write_all(help_text.as_bytes());
         return;
+    }
+
+    if let Err(error) = paged.write_str_flush() {
+        debug!("failed to flush pager: {error}");
+        // Don't write to stdout if flushing fails, because the text probably
+        // got written out.
     }
 
     paged.finalize();

--- a/nextest-runner/src/pager.rs
+++ b/nextest-runner/src/pager.rs
@@ -157,26 +157,6 @@ impl PagedOutput {
         }
     }
 
-    /// Returns a writer for stdout.
-    ///
-    /// For terminal output, this returns stdout directly.
-    /// For paged output, this returns the pipe to the pager.
-    ///
-    /// # Panics
-    ///
-    /// Panics if called after [`finalize`](Self::finalize).
-    pub fn stdout(&mut self) -> &mut dyn Write {
-        match self {
-            Self::Terminal { stdout, .. } => stdout,
-            Self::ExternalPager { child_stdin, .. } => {
-                child_stdin.as_mut().expect("stdout called after finalize")
-            }
-            Self::BuiltinPager { out_writer, .. } => {
-                out_writer.as_mut().expect("stdout called after finalize")
-            }
-        }
-    }
-
     /// Returns true if output will be displayed interactively.
     ///
     /// This is used to determine whether to use human-readable formatting
@@ -206,6 +186,24 @@ impl PagedOutput {
         self.finalize_inner();
     }
 
+    // ---
+    // Helper methods
+    // ---
+
+    // This is not made public: we want everyone to go through WriteStr, which
+    // squelches BrokenPipe errors.
+    fn stdout(&mut self) -> &mut dyn Write {
+        match self {
+            Self::Terminal { stdout, .. } => stdout,
+            Self::ExternalPager { child_stdin, .. } => child_stdin
+                .as_mut()
+                .expect("stdout should not be called after finalize"),
+            Self::BuiltinPager { out_writer, .. } => out_writer
+                .as_mut()
+                .expect("stdout should not be called after finalize"),
+        }
+    }
+
     fn finalize_inner(&mut self) {
         match self {
             Self::Terminal { .. } => {
@@ -220,11 +218,14 @@ impl PagedOutput {
                 // Close stdin to signal EOF to the pager.
                 drop(stdin);
 
-                // Wait for the pager to exit.
-                if let Err(error) = child.wait() {
+                // Wait for the pager to exit. (Ignore broken pipes -- they're
+                // expected with less.)
+                if let Err(error) = child.wait()
+                    && error.kind() != io::ErrorKind::BrokenPipe
+                {
                     warn!("failed to wait on pager: {error}");
                 }
-                // Note: We intentionally ignore the exit status. The pager may
+                // Note: We intentionally ignore the exit status from the child process. The pager may
                 // exit with a non-zero status if the user quits early (e.g.,
                 // pressing 'q' in less), which is normal behavior.
             }
@@ -279,11 +280,19 @@ impl Drop for PagedOutput {
 
 impl WriteStr for PagedOutput {
     fn write_str(&mut self, s: &str) -> io::Result<()> {
-        self.stdout().write_all(s.as_bytes())
+        squelch_broken_pipe(self.stdout().write_all(s.as_bytes()))
     }
 
     fn write_str_flush(&mut self) -> io::Result<()> {
-        self.stdout().flush()
+        squelch_broken_pipe(self.stdout().flush())
+    }
+}
+
+fn squelch_broken_pipe(res: io::Result<()>) -> io::Result<()> {
+    match res {
+        Ok(()) => Ok(()),
+        Err(e) if e.kind() == io::ErrorKind::BrokenPipe => Ok(()),
+        Err(e) => Err(e),
     }
 }
 
@@ -387,5 +396,38 @@ mod tests {
         output.finalize_inner();
         output.finalize_inner();
         // Drop will also try to finalize, should be safe.
+    }
+
+    #[test]
+    #[cfg(unix)]
+    fn test_external_pager_early_exit_squelches_broken_pipe() {
+        // `true` exits immediately, causing writes to fail with BrokenPipe.
+        let mut child = std::process::Command::new("true")
+            .stdin(Stdio::piped())
+            .spawn()
+            .expect("failed to spawn true");
+
+        let child_stdin = child.stdin.take().expect("stdin should be piped");
+
+        // Wait for the process to exit before constructing PagedOutput.
+        let _ = child.wait();
+
+        let mut output = PagedOutput::ExternalPager {
+            child,
+            child_stdin: Some(child_stdin),
+        };
+
+        output
+            .write_str("hello\n")
+            .expect("BrokenPipe should be squelched for write_str");
+        let error = output
+            .stdout()
+            .write(b"hello\n")
+            .expect_err("Write should fail with BrokenPipe");
+        assert_eq!(error.kind(), io::ErrorKind::BrokenPipe);
+        output
+            .write_str_flush()
+            .expect("BrokenPipe should be squelched for write_str_flush");
+        output.finalize();
     }
 }

